src/commonMain/kotlin/dev/romainguy/kotlin/math/Quaternion.kt

```diff
@@ -0,0 +1,332 @@
+        set(index1, v)
+        set(index2, v)
+        set(index3, v)
+    }
+
+    operator fun set(
+            index1: QuaternionComponent, index2: QuaternionComponent,
+            index3: QuaternionComponent, index4: QuaternionComponent, v: Float) {
+        set(index1, v)
+        set(index2, v)
+        set(index3, v)
+        set(index4, v)
+    }
+
+    operator fun unaryMinus() = Quaternion(-x, -y, -z, -w)
+
+    inline operator fun plus(v: Float) = Quaternion(x + v, y + v, z + v, w + v)
+    inline operator fun minus(v: Float) = Quaternion(x - v, y - v, z - v, w - v)
+    inline operator fun times(v: Float) = Quaternion(x * v, y * v, z * v, w * v)
+    inline operator fun div(v: Float) = Quaternion(x / v, y / v, z / v, w / v)
+
+    inline operator fun times(v: Float3) = (this * Quaternion(v, 0.0f) * inverse(this)).xyz
+
+    inline operator fun plus(q: Quaternion) = Quaternion(x + q.x, y + q.y, z + q.z, w + q.w)
+    inline operator fun minus(q: Quaternion) = Quaternion(x - q.x, y - q.y, z - q.z, w - q.w)
+    inline operator fun times(q: Quaternion) = Quaternion(
+            w * q.x + x * q.w + y * q.z - z * q.y,
+            w * q.y - x * q.z + y * q.w + z * q.x,
+            w * q.z + x * q.y - y * q.x + z * q.w,
+            w * q.w - x * q.x - y * q.y - z * q.z)
+
+    inline fun transform(block: (Float) -> Float): Quaternion {
+        x = block(x)
+        y = block(y)
+        z = block(z)
+        w = block(w)
+        return this
+    }
+
+    fun toEulerAngle() = eulerAngles(this)
+
+    fun toFloatArray() = floatArrayOf(x, y, z, w)
+}
+
+inline operator fun Float.plus(q: Quaternion) = Quaternion(this + q.x, this + q.y, this + q.z, this + q.w)
+inline operator fun Float.minus(q: Quaternion) = Quaternion(this - q.x, this - q.y, this - q.z, this - q.w)
+inline operator fun Float.times(q: Quaternion) = Quaternion(this * q.x, this * q.y, this * q.z, this * q.w)
+inline operator fun Float.div(q: Quaternion) = Quaternion(this / q.x, this / q.y, this / q.z, this / q.w)
+
+inline fun abs(q: Quaternion) = Quaternion(abs(q.x), abs(q.y), abs(q.z), abs(q.w))
+inline fun length(q: Quaternion) = sqrt(q.x * q.x + q.y * q.y + q.z * q.z + q.w * q.w)
+inline fun length2(q: Quaternion) = q.x * q.x + q.y * q.y + q.z * q.z + q.w * q.w
+inline fun dot(a: Quaternion, b: Quaternion) = a.x * b.x + a.y * b.y + a.z * b.z + a.w * b.w
+
+/**
+ * Rescale the Quaternion to the unit length
+ */
+fun normalize(q: Quaternion): Quaternion {
+    val l = 1.0f / length(q)
+    return Quaternion(q.x * l, q.y * l, q.z * l, q.w * l)
+}
+
+fun conjugate(q: Quaternion): Quaternion = Quaternion(q.w, -q.x, -q.y, -q.z)
+
+fun inverse(q: Quaternion): Quaternion {
+    val d = 1.0f / dot(q, q)
+    return Quaternion(q.w * d, -q.x * d, -q.y * d, -q.w * d)
+}
+
+fun real(q: Quaternion) = q.w
+
+fun imaginary(q: Quaternion) = q.xyz
+
+fun unreal(q: Quaternion) = Quaternion(q.x, q.y, q.z, 0.0f)
```

Comment:
Is this actually useful?

Reply:
No. I added it just because I like its name :smile: 
I will remove it.

Reply:
Fixed https://github.com/romainguy/kotlin-math/pull/33/commits/e60495ba84a929cbda6ede258a067ad5baa72d13